### DISCUSSION TITLE
Understanding cache file in opcua

### DISCUSSION THREAD

**Discussion by vidyarao123** [FreeOpcUa/python-opcua#1270] [Q&A]
Hello everyone, 

I am trying to work on this cache file of opcua. I have a few queries though

I went through the issue mentioned here in the link https://github.com/FreeOpcUa/python-opcua/issues/429 . Did not find an answer there. My code is as below

```
server = Server(shelffile="cache")
url = "opc.tcp://127.0.0.1:4840"
server.set_endpoint(url)
server.start()
XML = server.import_xml("output.xml")
obj = server.get_objects_node()

dump= address_space.AddressSpace.dump(node,'C:/Users/L_435239/eclipse-workspace/OPCUA_READ_WRT')
print (dump)

#load = address_space.AddressSpace.load(cache,'C:/Users/L_435239/eclipse-workspace/OPCUA_READ_WRT')
#print (load)

```
I am trying to achieve here that only for the first time, the variables / nodes shall be imported from the xml. But in the subsequent times, when the script is rerun, cache file must be used to load the nodes. 

My queries are 

1. When this line is executed, "server = Server(shelffile="cache")" , 3 files are being generated such as cache.bak, cache.dir, cache.dat . What I read in the issues column was , it generates .db file. So the files that are being generated in my system, are they correct ? 
2. I am using the dump () and load () here . But I am getting an error for these statements.  I would like to know if the format I have used is correct.

```
Traceback (most recent call last):
  File "C:\Users\L_435239\eclipse-workspace\OPCUA_READ_WRT\Python_memcached.py", line 68, in <module>
    dump= address_space.AddressSpace.dump(obj,'C:/Users/L_435239/eclipse-workspace/OPCUA_READ_WRT')
  File "C:\Program Files (x86)\Python38-32\lib\site-packages\opcua\server\address_space.py", line 569, in dump
    for nodeid, ndata in self._nodes.items():
AttributeError: 'Node' object has no attribute '_nodes'
```

Your help is much appreciated . Thanks

**Comment by zerox1212** [comment]
Read the doc strings. the `shelf` only works for caching the standard address space. It will not save your custom nodes from `output.xml`.

Again from the doc strings... you can't use `dump` and `load` and the shelf at the same time.

Stop using the shelf, then you can dump and load to persist your address space.

**Reply by vidyarao123** [reply]
Alright , But I don't quite understand the error here. 

```
Traceback (most recent call last):
  File "C:\Users\L_435239\eclipse-workspace\OPCUA_READ_WRT\Python_memcached.py", line 76, in <module>
    dump = address_space.AddressSpace.make_aspace_shelf(obj,'C:/Users/L_435239/eclipse-workspace/OPCUA_READ_WRT')
  File "C:\Program Files (x86)\Python38-32\lib\site-packages\opcua\server\address_space.py", line 594, in make_aspace_shelf
    for nodeid, ndata in self._nodes.items():
AttributeError: 'Node' object has no attribute '_nodes'
```

Could you please help me understand the error ?

**Reply by zerox1212** [reply]
You are calling `dump` on the `LazyLoadingDict` created by the shelf and it doesn't have an iterable or something along those lines. Set break points and step through the code if you want to find out. I don't remember exactly why this won't work, I just know doc strings have warnings about this that user should follow (I wrote them).

**Comment by vidyarao123** [comment]
Hi @zerox1212 , loading and dumping to .bin file is working now. I used dump_address_space() and load_address_space(). Along with the address space it saves the nodes also . Anyway my purpose is solved 

Thank you
